airbyte-integrations/connectors/source-mailgun/source_mailgun/source.py

```diff
@@ -0,0 +1,178 @@
+        """
+        :param response: the most recent response from the API
+        :return If there is another page in the result, a mapping (e.g: dict) containing information needed to query the next page in the response.
+                If there are no more pages in the result, return None.
+        """
+        next_page: Optional[HttpUrl] = response.json().get('paging', {}).get('next')
+        return {"url": next_page} if next_page and self._pre_parse_response(response) else None
+
+    def path(
+            self,
+            stream_state: Mapping[str, Any] = None,
+            stream_slice: Mapping[str, Any] = None,
+            next_page_token: Optional[Mapping[str, HttpUrl]] = None,
+    ) -> str:
+        # TODO: Requires improve URL creation in the CDK by using urllib.parse.urljoin
+        #  (airbyte_cdk.sources.streams.http.http._create_prepared_request)
+        return next_page_token["url"] if next_page_token else ""
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        yield from self._pre_parse_response(response)
+
+    @staticmethod
+    def _pre_parse_response(response: requests.Response) -> List:
+        return response.json()['items']
+
+
+class Domains(MailgunStream):
+    primary_key = "name"
+
+    def path(self, *args, next_page_token: Optional[Mapping[str, Any]] = None, **kwargs) -> str:
+        # return super().path(*args, **kwargs) or "domains"  # TODO: Requires the CDK update (see MailgunStream.path())
+        path: str = "domains"
+        if next_page_token:
+            token: str = next_page_token['url'].rpartition('/')[2]
+            path = f"{path}/{token}"
+
+        return path
+
+
+# Basic incremental stream
+class IncrementalMailgunStream(MailgunStream, ABC):
+
+    @staticmethod
+    def chunk_timestamps_range(start_timestamp: float, interval: Number = None) -> Iterable[Tuple[float]]:
+        """
+        Yield a tuple of beginning and ending timestamps of each day between the start timestamp and now.
+        """
+        interval = interval or 60 * 60 * 24  # 1 day
+        now = time.time()
+        if start_timestamp > now:
+            yield start_timestamp, start_timestamp
+
+        while start_timestamp <= now:
+            end_timestamp = start_timestamp + interval
+            yield start_timestamp, end_timestamp
+            start_timestamp = end_timestamp
+
+
+class Events(IncrementalMailgunStream):
+    # TODO: Event Polling. See https://documentation.mailgun.com/en/latest/api-events.html#event-polling
+
+    cursor_field = "timestamp"
+
+    primary_key = "id"
+
+    def __init__(self, config: Mapping[str, Any], *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Messages are stored for 3 days, so it prevents occasional attempt to read from the start of the Epoch
+        default_shift = 60 * 60 * 24 * 3
```

Comment:
the main problem with places like this and default values is that such behavior is declared to the user, that is why it is better to have all defaults for configuration declared on the level of schema.
Because you are not using Pydantic, this will be just an explanation (description) in the spec and additional logic in the source (probably, as you don't want this to be duplication in every incremental stream?)

Reply:
Added to the spec description.